src/ProtocolFeeController.sol

```diff
@@ -0,0 +1,117 @@
+    ///   2. a portion of the total fee based on 'protocolFeeSplitRatio' for static fee pools
+    mapping(PoolId => uint24) public defaultProtocolFees;
+
+    address public immutable poolManager;
+
+    /// @notice emit when the protocol fee is collected
+    event ProtocolFeeCollected(Currency indexed currency, uint256 amount);
+
+    constructor(address _clPoolManager) Ownable(msg.sender) {
+        poolManager = _clPoolManager;
+    }
+
+    /// @notice Set the ratio of the protocol fee in the total fee
+    function setProtocolFeeSplitRatio(uint256 newProtocolFeeSplitRatio) external onlyOwner {
+        if (newProtocolFeeSplitRatio > ONE_HUNDRED_PERCENT_RATIO) revert InvliadProtocolFeeSplitRatio();
+        protocolFeeSplitRatio = newProtocolFeeSplitRatio;
+    }
+
+    function setDefaultProtocolFee(PoolKey memory key, uint24 newProtocolFee) external onlyOwner {
+        if (address(key.poolManager) != poolManager) revert InvalidPoolManager();
+        /// @notice Validate the protocol fee, revert if fee is over the limit
+        if (!newProtocolFee.validate()) revert IProtocolFees.ProtocolFeeTooLarge(newProtocolFee);
+
+        defaultProtocolFees[key.toId()] = newProtocolFee;
+    }
+
+    /// @notice Get the protocol fee for a pool given the conditions of this contract
+    /// @return protocolFee The pool's protocol fee, expressed in hundredths of a bip. The upper 12 bits are for 1->0
+    function protocolFeeForPool(PoolKey memory poolKey) external view override returns (uint24 protocolFee) {
+        if (address(poolKey.poolManager) != poolManager) revert InvalidPoolManager();
+
+        // in case we've set a default protocol fee for the pool
+        uint24 defaultProtocolFee = defaultProtocolFees[poolKey.toId()];
+        if (defaultProtocolFee != 0) {
+            // already bi-directional fee so that it can be different in each direction
+            return defaultProtocolFee;
+        }
+
+        // otherwise, calculate the protocol fee based on the predefined rule
+        uint256 lpFee = poolKey.fee;
+        if (lpFee == LPFeeLibrary.DYNAMIC_FEE_FLAG) {
+            /// @notice for dynamic fee pools, the default protocol fee is 0
+            return _buildProtocolFee(0);
+        } else {
+            /// @notice for static fee pools, the protocol fee should be a portion of the total fee based on 'protocolFeeSplitRatio'
+            /// @dev the formula is derived from the following equation:
+            /// totalSwapFee = protocolFee + (1 - protocolFee) * lpFee = protocolFee / protocolFeeSplitRatio
+            uint24 oneDirectionProtocolFee = uint24(
+                lpFee * ONE_HUNDRED_PERCENT_RATIO
+                    / (
+                        lpFee + ONE_HUNDRED_PERCENT_RATIO * ONE_HUNDRED_PERCENT_RATIO / protocolFeeSplitRatio
+                            - ONE_HUNDRED_PERCENT_RATIO
+                    )
+            );
+
+            // cap the protocol fee at 0.4%, if it's over the limit we set it to the max
```

Comment:
When writing the test cases i found something unexpected:

Since protocol fee is capped by 0.4% which means only when lpFee is under roughly 0.82% that rule "protocolFee = 33% * totalFee " can hold.

Saying key.fee = 1%, then protocolFee will still be 0.4%, in this case the ratio being distributed to the protocol will under 33%, is this what we want ?

Reply:
Won't solve in this PR